### ISSUE TITLE
Discussion: async policy handling

### ISSUE THREAD

**Issue by TomKaltz** [Rotorsoft/eventually-monorepo#78]
Since policy processing is handled serially from the all stream, there is a possibility for a policy to severely lag behind if slow queries are used in the policy handler.  My current solution is to have the policy add a job to a BullMQ queue using an external system artifact.  This results in a lot of boilerplate code just to make policy handling performant.

It dawned on me that this sort of async parallel policy handling might be a candidate for a core feature or plugin for eventually.  The general idea is that a policy artifact could have a configurable flag to allow queued/async handling instead of the outbox/serial/checkpoint style handling.  The subscription store will mark what events were successfully put on the queue system and the queue system would handle redelivery, dead lettering, etc.

Does this type of feature sound wildly outside of the scope of a framework like eventually?  Is there a huge gotcha I might be missing?

**Comment by Rotorsoft** [comment]
This is a topic I was planning to tackle next, but haven't found the time to focus on it. It's basically the concept of subscriptions by correlation instead of at the artifact level, so you can only serialize flows that are strongly correlated. This should solve the scalability issues you are describing, but it requires some work at the broker level to understand new metadata to auto configure the system.

**Comment by TomKaltz** [comment]
...continuing from [here](https://github.com/Rotorsoft/eventually-monorepo/pull/89#issuecomment-2557205142), this is what is roaming around in my head.  It took a lot of iteration with Claude Sonnet to elucidate visually what I was thinking but here is how I envision the use of the new PostgresOrderedMessageQueue to allow scalable consumer processing.

Here are the important bits i'm trying to convey...
* broker enqueues events in strict order to each consumer's queue table based on it's event name interest
* each consumer subscription is acked as batches are enqueued
  * policy consumer queue is "streamed" by the original event stream id
  * process manager consumer queue is "streamed" by the actor name/hash (calculated by the broker using the artifact at time of enqueuing to the consumer's queue)
* Each consumer queue processor will dequeue and fill concurrent processing slots as messages are available. (2 in our examples but can be configured for more)
  * When a slot is available, the next ordered event(from an unlocked stream) is dequeued and begins processing
  * The entire stream/actor hash is locked during processing, this ensures ordered processing of each stream/actor hash
  * Events from the same stream/actor hash will never process at the same time...they will be processed one-by-one in strict order.  This is guaranteed by the dequeue implementation.
  * Streams processing does not necessarily stick to a certain slot, events are distributed as slots and new events are available (from unlocked streams).

I'm working on a POC broker right now that would handle all this.  So far I've only implemented policy event enqueueing.  I will build the processing side next and let you know if I can get something working.

Here are the diagrams that Claude Sonnet and I came up with.  I know the styling is messy, I apologize.

# Lets start with how a policy would be handled...
```mermaid
%%{init: {'theme': 'base', 'themeVariables': { 'lineColor': '#666666' }}}%%
graph TB
    subgraph EventStore["Event Store (Single Ordered Log)"]
        direction LR
        E1["1: Registered<br/>Stream1"] --> 
        E2["2: Arrived<br/>Stream2"] --> 
        E3["3: Registered<br/>Stream1"] --> 
        E4["4: Departed<br/>Stream3"] --> 
        E5["5: Registered<br/>Stream2"] --> 
        E6["6: Arrived<br/>Stream1"] -->
        E7["7: Departed<br/>Stream3"] -->
        E8["8: Arrived<br/>Stream2"] -->
        E9["9: Departed<br/>Stream1"] -->
        E10["10: Registered<br/>Stream3"]
    end

    subgraph Subscriptions["Consumer Subscriptions"]
        S1["Consumer1<br/>Checkpoint: 1<br/>Interested in: Arrived, Departed Events"]
    end

    subgraph Broker["Policy Broker"]
        B1["Batch Query Handler<br/>Fetches events after checkpoint<br/>Filters by event names"]
    end
    
    subgraph Queue1["Consumer1 Queue (Events in Event Store Order)"]
        direction LR
        Q1E1["2: Arrived<br/>Stream2"] --> 
        Q1E2["4: Departed<br/>Stream3"] --> 
        Q1E3["6: Arrived<br/>Stream1"] --> 
        Q1E4["7: Departed<br/>Stream3"] --> 
        Q1E5["8: Arrived<br/>Stream2"] --> 
        Q1E6["9: Departed<br/>Stream1"]
    end

    subgraph Processor["Parallel Event Processor<br/>(See Timeline Below)"]
        P1["Slot1"]
        P2["Slot2"]
    end

    %% Main flow connections
    EventStore --> |"Query events after checkpoint<br/>filtered by event names"| Broker
    Broker --> Queue1
    Queue1 --> Processor

    %% Subscription influence
    Subscriptions --> |"Checkpoint info"| Broker

    classDef store fill:#f9f,stroke:#333,stroke-width:2px,color:#000
    classDef sub fill:#bbf,stroke:#333,stroke-width:2px,color:#000
    classDef broker fill:#bfb,stroke:#333,stroke-width:2px,color:#000
    classDef queue fill:#ffb,stroke:#333,stroke-width:2px,color:#000
    classDef processor fill:#ffd,stroke:#333,stroke-width:2px,color:#000

    class EventStore store
    class Subscriptions sub
    class Broker broker
    class Queue1 queue
    class Processor processor
```

```mermaid
gantt
    title Slot-Based View: Processing Timeline
    dateFormat x
    axisFormat %L ms
    
    section Consumer1 Slot1
    E2 S2    :crit, 0, 30
    E6 S1    :done, 30, 40
    E7 S3    :active, 40, 50
    E9 S1    :done, 50, 80
    
    section Consumer1 Slot2
    E4 S3    :active, 0, 40
    E8 S2    :crit, 40, 70
```

# Now here is a diagram showing how eventually ProcessManagers would be handled...
```mermaid
graph TB
    subgraph EventStore["Event Store (Single Ordered Log)"]
        direction LR
        E1["1: Registered"] --> 
        E2["2: Arrived"] --> 
        E3["3: Registered"] --> 
        E4["4: Departed"] --> 
        E5["5: Registered"] --> 
        E6["6: Arrived"] -->
        E7["7: Departed"] -->
        E8["8: Arrived"] -->
        E9["9: Departed"] -->
        E10["10: Registered"]
    end

    subgraph Subscriptions["Process Manager Subscriptions"]
        S1["ProcessManager1<br/>Checkpoint: 1<br/>Interested in: Arrived, Departed Events<br/>Correlate by: actorHash"]
    end

    subgraph Broker["Process Manager Broker"]
        B1["Batch Query Handler<br/>Fetches events after checkpoint<br/>Filters by event names<br/>Groups by actorHash"]
    end
    
    subgraph Queue1["ProcessManager1 Queue (Events in actorHash Order)"]
        direction LR
        Q1E1["2: Arrived<br/>Stream: xyz"] --> 
        Q1E2["4: Departed<br/>Stream: xyz"] --> 
        Q1E3["6: Arrived<br/>Stream: abc"] --> 
        Q1E4["7: Departed<br/>Stream: def"] --> 
        Q1E5["8: Arrived<br/>Stream: ghi"] --> 
        Q1E6["9: Departed<br/>Stream: abc"]
    end

    subgraph Processor["Parallel Event Processor<br/>(See Timeline Below)"]
        P1["Slot1"]
        P2["Slot2"]
    end

    %% Main flow connections
    EventStore --> |"Query events after checkpoint<br/>filtered by event names"| Broker
    Broker --> Queue1
    Queue1 --> Processor

    %% Subscription influence
    Subscriptions --> |"Checkpoint info"| Broker

    classDef store fill:#f9f,stroke:#333,stroke-width:2px
    classDef sub fill:#bbf,stroke:#333,stroke-width:2px
    classDef broker fill:#bfb,stroke:#333,stroke-width:2px
    classDef queue fill:#ffb,stroke:#333,stroke-width:2px
    classDef processor fill:#ffd,stroke:#333,stroke-width:2px

    class EventStore store
    class Subscriptions sub
    class Broker broker
    class Queue1 queue
    class Processor processor
```

```mermaid
gantt
    title Slot-Based View: Processing Timeline
    dateFormat x
    axisFormat %L ms
    
    section ProcessManager1 Slot1
    E2 xyz    :crit, 0, 30
    E4 xyz    :crit, 30, 40
    E8 ghi    :active, 40, 60
    E9 abc    :done, 60, 80
    
    section ProcessManager1 Slot2
    E6 abc    :done, 0, 25
    E7 def    :active, 25, 100
```

**Comment by Rotorsoft** [comment]
This aligns closely with what I have in mind, but the main difference is that you introduce slots to support concurrent processing. In contrast, I’m envisioning this being resolved at the enqueueing level. Essentially, queues in the pool would behave like your slots, eliminating the need for explicit slot management.

Here’s how I’m visualizing the broker’s role:

### 1. Scheduling Events
The broker would schedule new events into “correlated queues” within a pool. Queues that fail to empty could be moved to a DLQ (Dead Letter Queue) facility and "released" back to the pool.

There will be basically two correlation types:

- Type 1: Aggregate Streams
For business flows fully encapsulated within an aggregate, where multiple aggregate instances are independent of each other, we can schedule their lifecycles in separate concurrent queues. For example, consider independent aggregates like user accounts that don’t interact among them. In this case, consumers (e.g., policies or projections) are correlated to individual aggregate instances, allowing parallel processing.

- Type 2: Process Managers, Sagas, and High-Level Projections
For business flows involving the coordination of multiple aggregate instances, where event order matters, correlation occurs at a higher level. For instance, if user 1 reserves a resource, user 2 shouldn’t consume it before the reservation completes. Here, the correlation key should be derived from "shared" attributes of the business flow (e.g., a resource ID as the unique identifier for the process instance). Events from multiple streams/aggregates are placed in a correlated queue for this ID and delivered sequentially to consumer policies. Concurrent processing would still be possible across different flow IDs.

### 2. Dispatching Events
Once queues are properly constructed, the broker’s role becomes simpler: dispatching events to consumers sequentially, handling retries, and managing DLQs (e.g., stopping a flow instance when necessary). This dispatching process should remain generic and straightforward.

## Basic Principles

The complexity of the broker lies primarily in the generalization of scheduling “correlated queues” from different streams based on computable correlation keys. This is the "business knowledge" that should be coded in the "subscription" schemas. Beyond that, the system should handle event dispatching with standard mechanisms like sequential delivery, retries, and queue management.

Of course, there is more to this, and additional features would include robust logging and tooling for visualizing the state of the queues, offering clear insights into system behavior.

Hope this makes sense.

**Comment by TomKaltz** [comment]
Where in your design concept are these queues processed?  I envisioned the consumers queues(one queue/table for each consumer artifact) being processed outside of the broker, by consumer queue processors that can be spun up as needed for parallel processing.  Again, the PostgresOrderedMessageQueue implementation allows for parallel dequeueing and will ensure serialized message processing within a stream("correlation key", I like this nomenclature!!) no matter how many consumer queue processor instances are competing for message.

I guess I ask the above question being I don't think I understand the concept or need for queue pools.  What immediately comes to my mind is that this would introduce a sort of sharding responsibility at the broker level that might not need to exist?  In my mind the broker should just enqueue each message to the appropriate consumer queue tables with either it's aggregate stream id or it's calculated correlation key.  The correlation key is calculated by calling artifact.actor[eventName](event).

This reminds me, we should probably change the nomenclature for the MessageQueue interface/column to "correlationKey" and not "stream" since we probably want to leave all the message data (like stream id) intact in the data column?

**Comment by Rotorsoft** [comment]
You seem to be placing the complexity in the queue processing side, where
concurrency is determined by creating slots. I’m thinking that it can be
placed in the scheduling side, when queues are already created with
concurrency in mind, one queue per process instance, that’s why you need a
pool to handle a large number of instances similar to web requests handled
by a thread pool. We need to think about the trade offs of each approach,
but the results are the same.

**Comment by TomKaltz** [comment]
You could remove the concept of slots inside a queue's processor and just make the processor simply dequeue and process one message at a time in a loop.  Then you can spin up as many instances as you need based on demand and it accomplishes the same exact thing.  The PostgresOrderedMessageQueue implementation guards for correlationKey concurrency by not allowing any of that queue's multiple producer processor instances to dequeue/process a given correlationKey message at the same time.  Lets say processor1 is processing a message with correllationKey1 and the next message in the queue also has correllationKey1, when another processor instance calls dequeue while processor1 is processing correllationKey1, the dequeue implementation dequeues the next available message that doesn't have that correllationKey (since that stream is locked in the database).

To me this reduces complexity because the broker isn't concerned with sharding into a queue pool.  The messages get queued in each consumer's queue table, which provides logical organization and containment for those messages.  The user can choose how many processor instances to spin up to process that queue table and the PostgresOrderedMessageQueue dequeue implementation already guarantees that correlated messages will be processed in strict order...even when multiple processes are dequeueing from that queue.

**Comment by Rotorsoft** [comment]
I need to think more about this with a clear mind.

1. We have business flows implemented via eventually artifacts akin to
classes in OOP.

2. Flows are implemented using a combination of aggregates and policies and
have a finite lifetime (ideally)…  policies are the automations and can
span multiple aggregates (sagas, etc)

3. Each flow can have many (potentially concurrent) uncorrelated instances
(recorded as correlated streams of events) akin to object instances in OOP

4. Instances have a correlation key and can be reduced to states
(projected) either as a contained aggregate or using projections for
viewing or reference

Basically you are proposing one queue per consumer artifact (kind of the
existing broker) that gets resolved later, and I’m thinking about one front
on-demand queue per flow instance. The limitation of the existing broker is
that it serializes at the consumer artifact level.

One flow can be composed of many consumer artifacts. (This is important
when evaluating a queue per consumer approach), while a flow instance is
unique.

Queues are ephemeral memory that gets allocated on demand when new events
arrive (in real time or when replaying) and their main responsibility is to
serialize correlated events to a flow instance.

We need to keep a watermark on events processed by instance in order to
support parallel processing. That can be a large table of watermarks
identifying all instances in the system (need to think about this)

Since we can have a large amount of active flows in the system, we need a
pool of resources when processing.

First question is:
Is it better to schedule (enqueue) instances upfront or schedule consumers
classes and delegate concurrent instances to the dispatcher?

Probably comes down to how we define subscriptions and manage watermarks.
The internal implementation can vary.

I’m learning for a queue per instance to keep the queue adapters very
simple, otherwise we have to deal with correlations inside the queue.

**Comment by TomKaltz** [comment]
However this gets implemented, it's clear to me that each event handling artifact will need to be provided functions that calculate it's correllationKey based on the event it's processing, exactly like how "actor" is specified in an eventually ProcessManager, right?

**Comment by Rotorsoft** [comment]
Not necessarily if the queue is already correlated and it’s guaranteed to
deliver the right events all the time.

The key question is where to place the responsibility to serialize
correlated events

**Comment by Rotorsoft** [comment]
I'll wait for your broker implementation and will provide feedback, but I think you are on the right track. 

If the broker you have in mind takes care of enqueuing events by correlation key in your ordered msg queue, that will represent the pool of queues. The other side is already taking care of delivering concurrently uncorrelated flows, and just needs some basic reliability guarantees and flagging the ones that are blocked by errors independently of the rest (DLQ like)

Looking forward to seeing that work 👍

**Comment by TomKaltz** [comment]
@Rotorsoft give this a whirl https://github.com/TomKaltz/ConsumerQueueExp

**Comment by Rotorsoft** [comment]
It works on my end and looks fine at first glance 👍 

The end goal I had in mind was to refactor the broker interface and hide the scheduling/dispatching (including subscriptions) behind the bootstrapping code, so no more public subscription stores or queue stores... we can probably make it work with just a single subscriptions table and a single message queue per service/process.

The other thing I noticed is that you are still using the old subscriptions interface and some of the existing broker logic, but with the introduction of correlated queues, the broker implementation should be much simpler.

Just high level without thinking much about the details:
- bootstrap broker with one queue adapter per service
- listening/polling for new events to be enqueued by correlation key (put on top of dispatcher priority queue)
- should probably store watermarks by correlation key too
- the generic broker loop can build a small priority queue in memory (pool) to keep track of active correlation queues and push until empty ...  feels like this should be a lot simpler than the current implementation... probably needs some sketching too for clarity, but in any case, I was never happy with the existing broker.

**Comment by TomKaltz** [comment]
Yeah I was only handling policies...this was just hacked together to get it to work and to elucidate what I was thinking.  There are still a ton of concepts and bugs that need to be worked out.

In my design idea/concept, the current subscription model seems to still work great.  The event store is written to atomically then it kicks off the broker to dispatch events to each individual consumer queue....the subscription watermark tell us which messages were successfully dispatched to their appropriate consumer queues from the all stream.  The consumer queue tables shows which correllationKeys are unprocessed (inverse of a correllationKey watermark table, which could grow forever).  We could query with a partial index, probably) to get a list of in-flight/unprocessed keys.  In my mind the above promotes the best durability without having to store correllationKey watermarks in perpetuuty.

**Comment by Rotorsoft** [comment]
That's true, but imagine the amount of value you can extract from that system level projection. The stats feature is another incomplete area that would benefit from a table with all instances (aggregates and PMs) including their class, watermark, etc

**Comment by TomKaltz** [comment]
Ah yes, [Marten](https://martendb.io/events/storage.html#database-tables) does this.  I like the idea.

**Comment by TomKaltz** [comment]
@Rotorsoft I've completely redesigned my POC according to what you've been talking about.  Give it a try if you have some free time... https://github.com/TomKaltz/ConsumerQueueExp.  Readme has the details.

NodeJS really shines in the concurrency department here.  I was able to keep increasing processing concurrency at seemingly no cost.  More tuning and testing is needed but I'm very impressed at how postgres handles these workloads.

**Comment by Rotorsoft** [comment]
Hey @TomKaltz , this looks great! I tested it locally, and the performance seems to meet the needs of most business scenarios I’m familiar with. 

Here are a few thoughts:

- As we discussed, polling remote event streams is essential for supporting fully distributed systems where each bounded context is physically separated and implemented as a microservice. In that context, sharing the connection pool for extra performance gain feels like a localized optimization and might not be as critical.
- I like how the ConsumerQueueingBroker adapter encapsulates the complexity you’re aiming for, especially the ordered, concurrent processing of uncorrelated streams. While specific creation options for each adapter are fine, I think there’s some shared functionality (such as seeding, starting, pausing, stopping, etc) that belongs in the broker’s port interface. The broker is the "engine of the system", so having a simple in-memory implementation of this interface for unit testing and better abstraction understanding would be valuable alongside more complex implementations like yours.
- Let’s discuss the subscription model/interface further. We need to ensure it supports integrating remote and local streams through a straightforward declaration of the protocol and the address to listen and poll for new events (Like listening/polling from REST endpoints)
- There’s quite a bit to unpack in the StreamConsumerProcessor implementation, which is the core of this POC. We should explore ways to map correlation key generation declaratively at the handler level. This would allow us to maintain a consistent and predictable flow identifier at the logical model level, regardless of the broker or stream processor adapter being used.
- I think we could break down the responsibilities at the processing level into more modular components and interfaces, such as:
  - Listening/Polling Engine: Responsible for managing consumer registration and mapping, as well as lifecycle operations like start, pause, and stop.
  - Stream Watermarks and Persistence: Focused on handling stream watermarks and correlation key persistence (e.g., seeding and tracking state).
  - Error Handling: Dedicated to managing retries, backoff strategies, and dead-letter queues (DLQ).
- These refinements would promote better reusability and composability of core responsibilities, turning them into framework-level building blocks.
- There are aspects like processing state_event commits that seem out of scope for the broker itself. These might be better placed elsewhere in the system where they align more closely with their logical purpose.

This is just a first pass at the implementation and my thoughts on integrating this POC into the main repo. I’ll likely have more points to discuss as I dive deeper into the details.

Again, thanks so much for putting this together! I hope these comments are helpful.
